### ISSUE TITLE
@return mixed[]|null not converted to : ?array

### ISSUE THREAD

**Issue by Soullivaneuh** [slevomat/coding-standard#247]
The following code part:

```php
/**
 * @return mixed[]|null
 */
public function getStats(Uptime $uptime, DateTime $start, ?string $type = null)
```

Should be converted to:

```php
/**
 * @return mixed[]|null
 */
public function getStats(Uptime $uptime, DateTime $start, ?string $type = null): ?array
```

Am I right?

Using last stable version. Configuration:

```xml
<?xml version="1.0"?>
<ruleset name="FlintCI Standard">
    <file>src</file>
    <file>src-dev</file>
    <file>tests</file>

    <exclude-pattern>tests/fixtures/repos</exclude-pattern>

    <rule ref="Symfony">
        <exclude name="Symfony.Commenting.License"/>
        <exclude name="Symfony.Commenting.Annotations.Invalid"/>
        <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/120 -->
        <exclude name="Symfony.Commenting.FunctionComment"/>
        <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/119 -->
        <exclude name="Symfony.Functions.ReturnType.Invalid"/>
        <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/27 -->
        <exclude name="PEAR.Functions.FunctionCallSignature"/>
        <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/44 -->
        <exclude name="Squiz.Strings.ConcatenationSpacing"/>
        <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/122 -->
        <exclude name="Squiz.Functions.MultiLineFunctionDeclaration.ContentAfterBrace"/>
        <exclude name="Squiz.WhiteSpace.ScopeClosingBrace.ContentBefore"/>
        <exclude name="PEAR.WhiteSpace.ScopeClosingBrace.Line"/>
        <exclude name="Symfony.Formatting.BlankLineBeforeReturn.Invalid"/>
    </rule>

    <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/121 -->
    <rule ref="Symfony.NamingConventions.ValidClassName.InvalidAbstractName">
        <exclude-pattern>*TestCase.php</exclude-pattern>
    </rule>

    <rule ref="SlevomatCodingStandard">
        <exclude name="SlevomatCodingStandard.ControlStructures.DisallowYodaComparison"/>
        <exclude name="SlevomatCodingStandard.Namespaces.FullyQualifiedClassNameInAnnotation"/>
        <!-- @see https://github.com/slevomat/coding-standard/issues/245 -->
        <exclude name="SlevomatCodingStandard.Namespaces.UseOnlyWhitelistedNamespaces"/>
        <!-- @see https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3471 -->
        <exclude name="SlevomatCodingStandard.Namespaces.ReferenceUsedNamesOnly.ReferenceViaFullyQualifiedNameWithoutNamespace"/>
        <!-- @see https://github.com/slevomat/coding-standard/issues/246 -->
        <exclude name="SlevomatCodingStandard.Namespaces.FullyQualifiedExceptions.NonFullyQualifiedException"/>
    </rule>
    <rule ref="SlevomatCodingStandard.Files.TypeNameMatchesFileName">
        <properties>
            <property
                name="rootNamespaces"
                type="array"
                value="
                    tests=>Tests,
                    src-dev=>Dev,
                "
            />
            <property name="ignoredNamespaces" type="array" value="DoctrineMigrations"/>
        </properties>
    </rule>
    <rule ref="SlevomatCodingStandard.Namespaces.ReferenceUsedNamesOnly">
        <properties>
            <property name="allowFullyQualifiedGlobalClasses" value="true"/>
            <property name="allowFullyQualifiedGlobalFunctions" value="true"/>
            <property name="allowFullyQualifiedGlobalConstants" value="true"/>
        </properties>
    </rule>
    <rule ref="SlevomatCodingStandard.Namespaces.UnusedUses">
        <properties>
            <property name="searchAnnotations" value="true"/>
        </properties>
    </rule>
    <rule ref="SlevomatCodingStandard.TypeHints.DeclareStrictTypes">
        <properties>
            <property name="newlinesCountBetweenOpenTagAndDeclare" value="2"/>
            <property name="spacesCountAroundEqualsSign" value="0"/>
        </properties>
    </rule>
    <rule ref="SlevomatCodingStandard.TypeHints.TypeHintDeclaration">
        <properties>
            <property
                name="usefulAnnotations"
                type="array"
                value="
                    @after,
                    @before,
                    @deprecated,
                    @required,
                    @ORM\,
                    @Route,
                    @Method,
                    @IsGranted
                "
            />
            <property name="enableEachParameterAndReturnInspection" value="true"/>
            <property name="enableObjectTypeHint" value="false"/>
        </properties>
    </rule>
    <rule ref="SlevomatCodingStandard.Types.EmptyLinesAroundTypeBraces">
        <properties>
            <property name="linesCountAfterOpeningBrace" value="0"/>
            <property name="linesCountBeforeClosingBrace" value="0"/>
        </properties>
    </rule>
</ruleset>
```

**Comment by kukulich** [comment]
No, `mixed` is required because when you use only `array` it's not possible to distinguish between "type hint is missing" and "type hint is mixed".

**Comment by Soullivaneuh** [comment]
@kukulich I'm not talking about the `mixed` phpdoc, it's indeed necessary to keep it.

I'm talking about not adding `: ?array`. Why should not be valid? We have array|null return and the PHPDoc for a more accurate return type.

Regards.

**Comment by kukulich** [comment]
Because “mixed[]” can be array, iterable, Iterator etc.
